File: api/routes/documents.py
import hashlib
import os
import re
import sys
import threading
import uuid
from pathlib import Path
from typing import Any

import io

import yaml
from PIL import Image
from fastapi import APIRouter, Body, Depends, HTTPException, UploadFile, File

ROOT = Path(__file__).parent.parent.parent
sys.path.insert(0, str(ROOT / "scripts"))

from classify_receipts import classify_filename, extract_with_ai_bytes  # noqa: E402
from api.auth import get_current_user, get_current_user_optional         # noqa: E402
from api.db import (                                                      # noqa: E402
    add_transaction, apply_dot_path_to_section,
    delete_document_record, file_already_applied,
    get_document_content, get_documents_for_user, init_db,
    upsert_document,
)

init_db()

router = APIRouter(tags=["documents"])

# job_id -> {status: "running"|"complete"|"error", extracted: dict|None, error: str|None}
_extract_jobs: dict[str, dict] = {}

ALLOWED_EXTENSIONS = {".pdf", ".jpg", ".jpeg", ".png", ".heic", ".tiff", ".csv"}

TAX_YEAR = 2025
MAX_UPLOAD_BYTES = 20 * 1024 * 1024  # 20 MB


_IMAGE_SUFFIXES = {".jpg", ".jpeg", ".png", ".heic", ".tiff"}
_MAX_DIMENSION = 1800
_JPEG_QUALITY = 85


def _compress_image(content: bytes) -> bytes | None:
    """Returns JPEG bytes on success, None if Pillow cannot decode the source."""
    try:
        img = Image.open(io.BytesIO(content))
        img = img.convert("RGB")
        if max(img.size) > _MAX_DIMENSION:
            img.thumbnail((_MAX_DIMENSION, _MAX_DIMENSION), Image.Resampling.LANCZOS)
        buf = io.BytesIO()
        img.save(buf, format="JPEG", quality=_JPEG_QUALITY, optimize=True)
        return buf.getvalue()
    except Exception:
        return None


def _safe_name(name: str) -> str:
    name = re.sub(r"[^\w.\-]", "_", name)
    return name[:200]


def _file_id(user_id: str, name: str, content: bytes) -> str:
    digest = hashlib.sha256(content).hexdigest()[:16]
    return hashlib.sha256(f"{user_id}:{name}:{digest}".encode()).hexdigest()[:12]


def _normalize(doc: dict) -> dict:
    """Shape DB row into the API response shape the frontend expects."""
    return {
        "file_id":    doc["id"],
        "file":       doc["filename"],
        "category":   doc.get("category", ""),
        "confidence": doc.get("confidence", ""),
        "size":       doc.get("size", 0),
        "note":       doc.get("note", ""),
        "extracted":  bool(doc.get("extracted", 0)),
        "uploaded_at": doc.get("uploaded_at", ""),
    }


# ── Upload / list / delete ─────────────────────────────────────────────────────

@router.post("/documents/upload")
def upload_document(file: UploadFile = File(...),
                    current_user: dict | None = Depends(get_current_user_optional)):
    suffix = Path(file.filename).suffix.lower()
    if suffix not in ALLOWED_EXTENSIONS:
        raise HTTPException(400, f"File type '{suffix}' not allowed.")

    safe = _safe_name(file.filename)
    content = file.file.read(MAX_UPLOAD_BYTES + 1)
    if len(content) > MAX_UPLOAD_BYTES:
        raise HTTPException(413, "File exceeds the 20 MB upload limit.")
    if suffix in _IMAGE_SUFFIXES:
        compressed = _compress_image(content)
        if compressed is not None:
            content = compressed
            safe = Path(safe).stem + ".jpg"
            suffix = ".jpg"
        elif suffix in {".heic", ".tiff"}:
            raise HTTPException(
                415,
                "HEIC/TIFF images must be converted to JPG/PNG/PDF before upload (image could not be decoded).",
            )
    fid = _file_id(uid, safe, content)

    info = classify_filename(safe, size=len(content))
    info["file_id"] = fid

    if current_user:
        upsert_document(
            current_user["id"], fid, safe,
            category=info.get("category", ""),
            confidence=info.get("confidence", ""),
            content=content,
            size=len(content),
            note=info.get("note", ""),
        )
    return info


@router.get("/documents")
def list_documents(current_user: dict | None = Depends(get_current_user_optional)):
    if not current_user:
        return {"files": []}
    docs = get_documents_for_user(current_user["id"])
    return {"files": [_normalize(d) for d in docs]}


@router.delete("/documents/{file_id}")
def delete_document(file_id: str,
                    current_user: dict | None = Depends(get_current_user_optional)):
    if not current_user:
        raise HTTPException(401, "Authentication required")
    deleted = delete_document_record(current_user["id"], file_id)
    if not deleted:
        raise HTTPException(404, f"Document '{file_id}' not found")
    return {"deleted": True, "file_id": file_id}


# ── AI Extraction ──────────────────────────────────────────────────────────────

def _run_extraction(job_id: str, content: bytes, filename: str) -> None:
    try:
        result = extract_with_ai_bytes(content, filename)
        _extract_jobs[job_id] = {"status": "complete", "extracted": result, "error": None}
    except Exception as exc:
        _extract_jobs[job_id] = {"status": "error", "extracted": None, "error": str(exc)}


@router.post("/documents/{file_id}/extract")
def extract_document(file_id: str,
                     current_user: dict | None = Depends(get_current_user_optional)):
    if not os.environ.get("ANTHROPIC_API_KEY"):
        raise HTTPException(503, "ANTHROPIC_API_KEY is not set")
    if not current_user:
        raise HTTPException(401, "Authentication required")
    content, filename = get_document_content(current_user["id"], file_id)
    if content is None:
        raise HTTPException(404, f"Document '{file_id}' not found")
    job_id = str(uuid.uuid4())
    _extract_jobs[job_id] = {"status": "running", "extracted": None, "error": None}
    threading.Thread(target=_run_extraction, args=(job_id, content, filename), daemon=True).start()
    return {"job_id": job_id}


@router.get("/documents/extract/{job_id}")
def extraction_status(job_id: str):
    job = _extract_jobs.get(job_id)
    if job is None:
        raise HTTPException(404, f"Extraction job '{job_id}' not found")
    return job


# ── Apply extracted updates ────────────────────────────────────────────────────

def _apply_dot_path(data: Any, dot_path: str, operation: str, value: Any) -> bool:
    """Traverse dot_path in data dict and apply set/add. Pure function — no I/O."""
    parts = dot_path.split(".")
    obj = data
    try:
        for part in parts[:-1]:
            if isinstance(obj, list):
                obj = obj[int(part)]
            elif isinstance(obj, dict):
                obj = obj.setdefault(part, {})
            else:
                return False
        last = parts[-1]
        if isinstance(obj, list):
            idx = int(last)
            while len(obj) <= idx:
                obj.append(None)
            if operation == "add":
                obj[idx] = round((obj[idx] or 0) + value, 2)
            else:
                obj[idx] = value
        elif isinstance(obj, dict):
            if operation == "add":
                obj[last] = round((obj.get(last) or 0) + value, 2)
            else:
                obj[last] = value
        else:
            return False
        return True
    except (KeyError, IndexError, TypeError, ValueError):
        return False


@router.post("/documents/apply")
def apply_extraction(body: Any = Body(...),
                     current_user: dict | None = Depends(get_current_user_optional)):
    """
    Body: {
      "meta": {file_id, filename, date, merchant, total_amount,
               deductible_pct, tax_category, benefit_ids, form_line},
      "updates": [{yaml_file|section, dot_path, operation, value, label}]
    }
    Legacy list format also accepted.
    """
    if isinstance(body, list):
        updates = body
        meta: dict = {}
    else:
        updates = body.get("updates", [])
        meta = body.get("meta", {})

    file_id = meta.get("file_id", "")
    deductible_pct = float(meta.get("deductible_pct", 1.0))
    deductible_pct = max(0.0, min(1.0, deductible_pct))
    user_id = current_user["id"] if current_user else None

    if file_id and user_id and file_already_applied(user_id, file_id):
        return {"applied": [], "skipped": [u.get("label", "") for u in updates], "duplicate": True}

    applied = []
    skipped = []

    for update in updates:
        section = update.get("yaml_file") or update.get("section", "")
        dot_path = update.get("dot_path", "")
        operation = update.get("operation", "set")
        raw_value = update.get("value")
        label = update.get("label", dot_path)

        if not section or not dot_path or raw_value is None:
            skipped.append(label)
            continue
        if ".." in section or "/" in section or "\\" in section:
            skipped.append(label)
            continue

        if operation == "add" and isinstance(raw_value, (int, float)) and raw_value > 0:
            value = round(raw_value * deductible_pct, 2)
        else:
            value = raw_value

        success = False
        if user_id:
            success = apply_dot_path_to_section(user_id, TAX_YEAR, section, dot_path, operation, value)
        else:
            yaml_path = ROOT / "user_data" / f"{section}.yaml"
            if yaml_path.exists():
                with open(yaml_path, encoding="utf-8") as f:
                    data = yaml.safe_load(f) or {}
                if _apply_dot_path(data, dot_path, operation, value):
                    with open(yaml_path, "w", encoding="utf-8") as f:
                        yaml.dump(data, f, default_flow_style=False, allow_unicode=True)
                    success = True

        if success:
            applied.append(label)
            if file_id and user_id:
                total_amount = float(raw_value) if isinstance(raw_value, (int, float)) else 0.0
                add_transaction({
                    "id": str(uuid.uuid4()),
                    "user_id": user_id,
                    "file_id": file_id,
                    "filename": meta.get("filename", ""),
                    "date": meta.get("date"),
                    "merchant": meta.get("merchant"),
                    "total_amount": total_amount,
                    "deductible_pct": deductible_pct,
                    "deductible_amount": round(total_amount * deductible_pct, 2),
                    "tax_category": meta.get("tax_category", ""),
                    "benefit_ids": meta.get("benefit_ids", []),
                    "form_line": meta.get("form_line", ""),
                    "section": section,
                    "dot_path": dot_path,
                    "status": "applied",
                    "label": label,
                })
        else:
            skipped.append(label)

    return {"applied": applied, "skipped": skipped, "duplicate": False}

File: backend-ts/src/routes/documents.ts
import fs from "node:fs";
import path from "node:path";
import crypto from "node:crypto";
import type { FastifyInstance, FastifyRequest } from "fastify";
import yaml from "js-yaml";
import { z } from "zod";
import { AppError } from "../lib/errors";
import { projectPaths } from "../lib/paths";
import { addTransaction, fileAlreadyApplied } from "../db/transactionsRepo";
import { applyDotPathToSection } from "../db/sectionRepo";
import { classifyFilename } from "../domain/documents/classifier";
import { extractWithAiBytes } from "../domain/documents/aiExecutor";
import {
  deleteDocumentRecord,
  getDocumentContent,
  getDocumentsForUser,
  saveDocumentExtraction,
  upsertDocument
} from "../db/documentsRepo";

const ALLOWED_EXTENSIONS = new Set([".pdf", ".jpg", ".jpeg", ".png", ".heic", ".tiff", ".csv"]);
const MAX_UPLOAD_BYTES = 20 * 1024 * 1024;

type MultipartLikeFile = {
  filename: string;
  toBuffer: () => Promise<Buffer>;
};

type ExtractionJobState = {
  status: "running" | "complete" | "error";
  extracted: Record<string, unknown> | null;
  error: string | null;
};

const extractionJobs = new Map<string, ExtractionJobState>();

const UploadPayloadSchema = z.object({
  filename: z.string().optional(),
  content: z.string().optional(),
  category: z.string().optional()
});

const FileParamsSchema = z.object({
  fileId: z.string().min(1)
});

const JobParamsSchema = z.object({
  jobId: z.string().min(1)
});

const UpdateInstructionSchema = z.object({
  label: z.string().max(200).optional(),
  section: z.string().regex(/^[A-Za-z0-9_]+$/).max(64).optional(),
  yaml_file: z.string().regex(/^[A-Za-z0-9_]+$/).max(64).optional(),
  dot_path: z.string().min(1).max(240),
  operation: z.enum(["set", "add"]).default("set"),
  value: z.unknown()
}).passthrough();

const ApplyMetaSchema = z.object({
  file_id: z.string().max(64).optional(),
  filename: z.string().max(200).optional(),
  date: z.string().regex(/^\d{4}-\d{2}-\d{2}$/).optional(),
  merchant: z.string().max(120).optional(),
  total_amount: z.number().finite().min(0).max(1_000_000_000).optional(),
  deductible_pct: z.number().finite().min(0).max(1).optional(),
  tax_category: z.string().max(80).optional(),
  benefit_ids: z.array(z.string().max(120)).max(50).optional(),
  form_line: z.string().max(80).optional()
}).passthrough();

const ApplyBodySchema = z.union([
  z.array(UpdateInstructionSchema),
  z.object({
    updates: z.array(UpdateInstructionSchema).optional(),
    meta: ApplyMetaSchema.optional()
  })
]);

function toBuffer(value: string | Buffer): Buffer {
  return Buffer.isBuffer(value) ? value : Buffer.from(value, "utf8");
}

function assertUploadSize(content: Buffer): void {
  if (content.length > MAX_UPLOAD_BYTES) {
    throw new AppError(413, "Uploaded file exceeds 20 MB limit");
  }
}

function safeName(name: string): string {
  return name.replace(/[^\w.-]/g, "_").slice(0, 200);
}

function fileId(userId: string, filename: string, content: Buffer): string {
  const digest = crypto.createHash("sha256").update(content).digest("hex").slice(0, 16);
  return crypto.createHash("sha256").update(`${userId}:${filename}:${digest}`).digest("hex").slice(0, 12);
}

function normalizeDocument(row: Awaited<ReturnType<typeof getDocumentsForUser>>[number]): Record<string, unknown> {
  return row;
}

function applyDotPath(data: Record<string, unknown>, dotPath: string, operation: string, value: unknown): boolean {
  const parts = dotPath.split(".");
  let current: unknown = data;

  try {
    for (const part of parts.slice(0, -1)) {
      if (Array.isArray(current)) {
        current = current[Number(part)];
      } else if (current && typeof current === "object") {
        const obj = current as Record<string, unknown>;
        if (!(part in obj) || obj[part] == null) {
          obj[part] = {};
        }
        current = obj[part];
      } else {
        return false;
      }
    }

    const last = parts[parts.length - 1];
    if (Array.isArray(current)) {
      const index = Number(last);
      while (current.length <= index) {
        current.push(null);
      }
      if (operation === "add") {
        current[index] = Number(current[index] ?? 0) + Number(value ?? 0);
      } else {
        current[index] = value;
      }
      return true;
    }

    if (current && typeof current === "object") {
      const obj = current as Record<string, unknown>;
      if (operation === "add") {
        obj[last] = Number(obj[last] ?? 0) + Number(value ?? 0);
      } else {
        obj[last] = value;
      }
      return true;
    }

    return false;
  } catch {
    return false;
  }
}

async function readUpload(request: FastifyRequest): Promise<{ filename: string; content: Buffer } | null> {
  const body = UploadPayloadSchema.optional().parse(request.body);
  if (body?.filename && body.content != null) {
    const content = toBuffer(body.content);
    assertUploadSize(content);
    return { filename: safeName(body.filename), content };
  }

  const multipartRequest = request as FastifyRequest & { file: () => Promise<MultipartLikeFile | undefined> };
  const file = await multipartRequest.file();
  if (!file) {
    return null;
  }
  const content = await file.toBuffer();
  assertUploadSize(content);
  return { filename: safeName(file.filename), content };
}

async function runExtraction(jobId: string, userId: string, fileId: string, filename: string, content: Buffer): Promise<void> {
  try {
    const extracted = await extractWithAiBytes(content, filename);
    await saveDocumentExtraction(userId, fileId, extracted);
    extractionJobs.set(jobId, { status: "complete", extracted, error: null });
  } catch (error) {
    extractionJobs.set(jobId, { status: "error", extracted: null, error: (error as Error).message });
  }
}

export async function registerDocumentsRoutes(app: FastifyInstance): Promise<void> {
  app.post("/documents/upload", { preHandler: app.authenticateOptional }, async (request) => {
    const uploaded = await readUpload(request);
    if (!uploaded) {
      throw new AppError(400, "No file uploaded");
    }

    const suffix = path.extname(uploaded.filename).toLowerCase();
    if (!ALLOWED_EXTENSIONS.has(suffix)) {
      throw new AppError(400, `File type '${suffix}' not allowed.`);
    }

    const userId = request.currentUser?.id ?? "";
    const id = fileId(userId, uploaded.filename, uploaded.content);
    const info = classifyFilename(uploaded.filename, uploaded.content.length);

    if (request.currentUser) {
      await upsertDocument(request.currentUser.id, id, uploaded.filename, uploaded.content, info);
    }

    return {
      file_id: id,
      file: uploaded.filename,
      category: info.category,
      confidence: info.confidence,
      size: uploaded.content.length,
      note: info.note,
      extracted: false,
      uploaded_at: new Date().toISOString()
    };
  });

  app.get("/documents", { preHandler: app.authenticateOptional }, async (request) => {
    if (!request.currentUser) {
      return { files: [] };
    }

    return { files: (await getDocumentsForUser(request.currentUser.id)).map(normalizeDocument) };
  });

  app.delete("/documents/:fileId", { preHandler: app.authenticate }, async (request) => {
    const userId = request.currentUser?.id;
    if (!userId) {
      throw new AppError(401, "Authentication required");
    }

    const { fileId } = FileParamsSchema.parse(request.params ?? {});
    const deleted = await deleteDocumentRecord(userId, fileId);
    if (!deleted) {
      throw new AppError(404, `Document '${fileId}' not found`);
    }

    return { deleted: true, file_id: fileId };
  });

  app.post("/documents/:fileId/extract", { preHandler: app.authenticateOptional }, async (request) => {
    if (!process.env.ANTHROPIC_API_KEY) {
      throw new AppError(503, "ANTHROPIC_API_KEY is not set");
    }

    const userId = request.currentUser?.id;
    if (!userId) {
      throw new AppError(401, "Authentication required");
    }

    const { fileId } = FileParamsSchema.parse(request.params ?? {});
    const doc = await getDocumentContent(userId, fileId);
    if (!doc.content) {
      throw new AppError(404, `Document '${fileId}' not found`);
    }

    const jobId = crypto.randomUUID();
    extractionJobs.set(jobId, { status: "running", extracted: null, error: null });
    queueMicrotask(() => {
      void runExtraction(jobId, userId, fileId, doc.filename, doc.content as Buffer);
    });
    return { job_id: jobId };
  });

  app.get("/documents/extract/:jobId", async (request) => {
    const { jobId } = JobParamsSchema.parse(request.params ?? {});
    const job = extractionJobs.get(jobId);
    if (!job) {
      throw new AppError(404, `Extraction job '${jobId}' not found`);
    }

    return job;
  });

  app.post("/documents/apply", { preHandler: app.authenticateOptional }, async (request) => {
    const body = ApplyBodySchema.parse(request.body ?? {});
    const updates = Array.isArray(body) ? body : (body.updates ?? []);
    const meta = Array.isArray(body) ? {} : (body.meta ?? {});

    const fileId = String(meta.file_id ?? "");
    const userId = request.currentUser?.id ?? null;
    const deductiblePct = Math.max(0, Math.min(1, Number(meta.deductible_pct ?? 1)));

    if (fileId && userId && await fileAlreadyApplied(userId, fileId)) {
      return { applied: [], skipped: updates.map((u) => String(u.label ?? "")), duplicate: true };
    }

    const applied: string[] = [];
    const skipped: string[] = [];

    for (const update of updates) {
      const section = String(update.yaml_file ?? update.section ?? "");
      const dotPath = String(update.dot_path ?? "");
      const operation = String(update.operation ?? "set");
      const rawValue = update.value;
      const label = String(update.label ?? dotPath);

      if (!section || !dotPath || rawValue == null || section.includes("..") || section.includes("/") || section.includes("\\")) {
        skipped.push(label);
        continue;
      }

      const value = operation === "add" && typeof rawValue === "number" && rawValue > 0
        ? Number((rawValue * deductiblePct).toFixed(2))
        : rawValue;

      let success = false;
      if (userId) {
        success = await applyDotPathToSection(userId, 2025, section, dotPath, operation, value);
      } else {
        const yamlPath = path.join(projectPaths.userData, `${section}.yaml`);
        if (fs.existsSync(yamlPath)) {
          const data = yaml.load(fs.readFileSync(yamlPath, "utf8")) as Record<string, unknown> | null;
          const mutable = (data && typeof data === "object" && !Array.isArray(data)) ? data : {};
          if (applyDotPath(mutable, dotPath, operation, value)) {
            fs.writeFileSync(yamlPath, yaml.dump(mutable, { noRefs: true, lineWidth: -1 }), "utf8");
            success = true;
          }
        }
      }

      if (success) {
        applied.push(label);
        if (fileId && userId) {
          const totalAmount = typeof rawValue === "number" ? rawValue : 0;
          await addTransaction({
            user_id: userId,
            file_id: fileId,
            filename: String(meta.filename ?? ""),
            date: meta.date ? String(meta.date) : "",
            merchant: meta.merchant ? String(meta.merchant) : "",
            total_amount: totalAmount,
            deductible_pct: deductiblePct,
            deductible_amount: Number((totalAmount * deductiblePct).toFixed(2)),
            tax_category: String(meta.tax_category ?? ""),
            benefit_ids: Array.isArray(meta.benefit_ids) ? meta.benefit_ids.map(String) : [],
            form_line: meta.form_line ? String(meta.form_line) : "",
            section,
            dot_path: dotPath,
            status: "applied",
            label
          });
        }
      } else {
        skipped.push(label);
      }
    }

    return { applied, skipped, duplicate: false };
  });
}
